### DISCUSSION TITLE
SRCError on release mode not present on debug

### DISCUSSION THREAD

**Discussion by ultrabug** [gklijs/schema_registry_converter#64] [Q&A]
Hi

- rustc 1.55.0 (c8dfcfe04 2021-09-06)

I read messages from rdkafka using this crate like this:

```rust
let decoded = match decoder.decode(m.payload()).await {
                    Ok(inner) => inner.value,
                    Err(err) => {
                        println!("failed to decode message: {:?}", err);
                        continue
                    },
```

When running on debug mode using `cargo run` I see no decode error message. But when I run the same code in release mode, I get a lot of:

```
SRCError { error: "Could not transform bytes using schema", cause: Some("Cannot convert i64 to usize: -53"), retriable: false, cached: false }
```

What would you recommend to find out what's wrong please?

Thanks!

**Comment by gklijs** [comment]
My first impression is that is was just accidental that is was based on a difference in debug and release mode.

It seems it read a faulty message, this might be a bug in the `avro-rs` library, but is impossible to say from the example.

Something you could do is print the value bytes, and get the schema from schema registry. And get a minimal example like in https://github.com/gklijs/schema_registry_converter/blob/95f48fbea5f3c6ed5fd3c937689517313eeb1702/src/async_impl/avro.rs#L593.

I can't really place the specific error, as it seems the schema (or at least `avro-rs`) wants to create an usize, and I don't understand why.

**Reply by ultrabug** [reply]
> My first impression is that is was just accidental that is was based on a difference in debug and release mode.

I get tons of them, always in release mode. If I reset my offset and run again but in debug mode, nothing.. so it's really about the release mode.

I'll try to dig into what you pointed out, I hope my (new) rust skills are enough to be up to this task... :)

Thank you

**Reply by gklijs** [reply]
I guess all go wrong, or all go right then? I don't have much experience with Rust, and never had other behavior on debug compared to release. So I'm afraid I can't really help you further. I'd love to hear what the problem was when you find out.

**Reply by ultrabug** [reply]
Not really, in release mode some messages are decoded okay, some not... very strange.

I'll surely get back to you here if I find out what's going on. For now the avro-rs repo has not been responsive to my low skilled questions :) let's hope this one raise more interest..

**Comment by ultrabug** [comment]
First finding: it's related to compilation optimizations

```
[profile.release]
opt-level = 0
```

Setting optimization level to 0 makes the problem go away. Any superior optimization will trigger the problem.
